### PR TITLE
Use beginBackgroundTask to protect DiskImageStore writes

### PR DESCRIPTION
**Pull request by darinwf**
Refactors BackgroundFileWriter so it can be used by DiskImageStore.

Issue #2838

### How was this tested?
- I updated existing automated tests

### CODE REVIEW

File: Client/Frontend/Browser/Tabs/SavedTab+ConfigureExtension.swift [hunk 1]
```diff
@@ -57,7 +57,7 @@ extension SavedTab {
             let imageStore = imageStore
         {
             tab.screenshotUUID = screenshotUUID
-            imageStore.get(screenshotUUID.uuidString) >>== { screenshot in
+            imageStore.get(screenshotUUID.uuidString) { screenshot in
```
Comment (chung-lai) [inline Client/Frontend/Browser/Tabs/SavedTab+ConfigureExtension.swift:60]: hmm...not sure what `>>==` operator is, is that we don't do any sql operation anymore for image?

File: Client/Frontend/Browser/Tabs/TabManagerStore.swift [hunk 2]
```diff
@@ -58,22 +58,19 @@ class TabManagerStore {
...
-                savedUUIDs.insert(screenshotUUID.uuidString)
-
-                imageStore?.put(screenshotUUID.uuidString, image: screenshot)
+            if let image = tab.screenshot, let uuid = tab.screenshotUUID {
+                screenshots.append(.init(key: uuid.uuidString, image: image))
             }
         }
-        // Clean up any screenshots that are no longer associated with a tab.
-        _ = imageStore?.clearExcluding(savedUUIDs)
+        imageStore?.putAll(screenshots)
```
Comment (chung-lai) [inline Client/Frontend/Browser/Tabs/TabManagerStore.swift:72]: nit: probably the method name should mention about it clears the unused screenshots

File: Storage/DiskImageStore.swift [hunk 3]
```diff
@@ -39,70 +32,97 @@ open class DiskImageStore {
...
+
+    /// Updates the store with the full set of entries that should be stored. Removes
+    /// any existing entries that are not included.
+    open func putAll(_ entries: [Entry]) {
+        processor.performTask {
+            var keysToKeep = Set<String>()
+
+            for entry in entries {
+                keysToKeep.insert(entry.key)
+                self.putOne(key: entry.key, image: entry.image)
```
Comment (chung-lai) [inline Storage/DiskImageStore.swift:83]: What do you think about logging here when an existing entry is found for debugging purpose?

Comment (chung-lai) [reply Storage/DiskImageStore.swift:83]: Or add logging when a file is successfully written?

**Review by chung-lai — APPROVED** [review]
Awesome refactoring! LGTM, left a few minor comments.

File: Client/Frontend/Browser/Tabs/SavedTab+ConfigureExtension.swift (hunk 1 above)
Comment (darinwf) [reply Client/Frontend/Browser/Tabs/SavedTab+ConfigureExtension.swift:60]: This is from `DeferredUtils.swift`. It just runs the callback only when there is success. I got rid of using `Deferred` and instead just made the second parameter to `DiskImageStore.get` be an ordinary callback function that gets passed a `UIImage?` that will be nil if there is a lookup error. I think this is a simpler API.

File: Client/Frontend/Browser/Tabs/TabManagerStore.swift (hunk 2 above)
Comment (darinwf) [reply Client/Frontend/Browser/Tabs/TabManagerStore.swift:72]: I was kind of intentionally burying the detail about cleanup. I'll give some more thought to the name....

File: Storage/DiskImageStore.swift (hunk 3 above)
Comment (darinwf) [reply Storage/DiskImageStore.swift:83]: I didn't feel a strong need for it and it can add noise to the log file when we are trying to study problems.

**Comment by chung-lai** [comment]
Looks like test fails to build.

**Pull request merged**